File: Data_Structures/Tree/Binary_Tree/Serialize/Serialize.cpp
#include <stdio.h> 
#define MARKER -1 

struct Node 
{ 
    int key; 
    struct Node* left, *right; 
}; 

Node* newNode(int key) 
{ 
    Node* temp = new Node; 
    temp->key = key; 
    temp->left = temp->right = NULL; 
    return (temp); 
} 
  
void serialize(Node *root, FILE *fp) 
{ 
    if (root == NULL) 
    { 
        fprintf(fp, "%d ", MARKER); 
        return; 
    } 
  
    fprintf(fp, "%d ", root->key); 
    serialize(root->left, fp); 
    serialize(root->right, fp); 
} 

int main() 
{ 
    struct Node *root        = newNode(32); 
    root->left               = newNode(12); 
    root->right              = newNode(23); 
    root->left->left         = newNode(40); 
    root->left->right        = newNode(33); 
    root->left->right->left  = newNode(50); 
    root->left->right->right = newNode(45); 
  
    FILE *fp = fopen("tree.txt", "w"); 
    if (fp == NULL) 
    { 
        puts("Error: Could not open file!"); 
        return 0; 
    } 
    serialize(root, fp); 
    fclose(fp);   
    return 0; 
} 

File: Data-Structures/Tree/Binary-Tree/Serialize/Serialize.py
class Node:
    def __init__(self, key, left=None, right=None):
        self.key = key
        self.left = left
        self.right = right

    def __repr__(self):
        left = f" left={self.left}" if self.left is not None else ""
        right = f" right={self.right}" if self.right is not None else ""
        return f"<Node object key={self.key}{left+right}>"


def serialize(root):
    if root is None:
        return "-1"

    # for each branch, return the key, the left then the right node
    output = (str(root.key), serialize(root.left), serialize(root.right))
    return " ".join(output)


if __name__ == "__main__":
    #     1
    #    / \
    #   2   3
    #  / \ / \
    # 4 -1 6  7
    root = Node(1, Node(2, Node(4)), Node(3, Node(6), Node(7)))
    print(root)

    with open("tree.txt", "w") as fp:
        fp.write(serialize(root))